File: src/main/java/net/frozenorb/potpvp/util/serialization/BlockVectorAdapter.java
package net.frozenorb.potpvp.util.serialization;

import com.google.gson.*;
import org.bukkit.util.BlockVector;

import java.lang.reflect.Type;

public class BlockVectorAdapter implements JsonDeserializer<BlockVector>, JsonSerializer<BlockVector> {

    @Override
    public BlockVector deserialize(JsonElement src, Type type, JsonDeserializationContext context) throws JsonParseException {
        return fromJson(src);
    }

    @Override
    public JsonElement serialize(BlockVector src, Type type, JsonSerializationContext context) {
        return toJson(src);
    }

    public static JsonObject toJson(BlockVector src) {
        if (src == null) {
            return null;
        }

        final JsonObject object = new JsonObject();

        object.addProperty("x", src.getX());
        object.addProperty("y", src.getY());
        object.addProperty("z", src.getZ());

        return object;
    }

    public static BlockVector fromJson(JsonElement src) {
        if (src == null || !src.isJsonObject()) {
            return null;
        }
        final JsonObject json = src.getAsJsonObject();

        final double x = json.get("x").getAsDouble();
        final double y = json.get("y").getAsDouble();
        final double z = json.get("z").getAsDouble();

        return new BlockVector(x, y, z);
    }

}

File: src/main/kotlin/net/frozenorb/potpvp/kt/util/serialization/BlockVectorAdapter.kt
package net.frozenorb.potpvp.kt.util.serialization

import com.google.gson.*
import org.bukkit.util.BlockVector
import java.lang.reflect.Type

class BlockVectorAdapter : JsonDeserializer<BlockVector>, JsonSerializer<BlockVector> {
    @Throws(JsonParseException::class)
    override fun deserialize(src: JsonElement, type: Type, context: JsonDeserializationContext): BlockVector? {
        return fromJson(src)
    }

    override fun serialize(src: BlockVector, type: Type, context: JsonSerializationContext): JsonElement {
        return toJson(src) as JsonElement
    }

    companion object {
        fun toJson(src: BlockVector?): JsonObject? {
            if (src == null) {
                return null
            }
            val `object` = JsonObject()
            `object`.addProperty("x", src.x as Number)
            `object`.addProperty("y", src.y as Number)
            `object`.addProperty("z", src.z as Number)
            return `object`
        }

        fun fromJson(src: JsonElement?): BlockVector? {
            if (src == null || !src.isJsonObject) {
                return null
            }
            val json = src.asJsonObject
            val x = json.get("x").asDouble
            val y = json.get("y").asDouble
            val z = json.get("z").asDouble
            return BlockVector(x, y, z)
        }
    }
}